### PR TITLE
Various optimizations

### PR DESCRIPTION
**Pull request by KalleZ**
This PR attempts to achieve a few things:

 * Minor optimizations by prefixing all global functions and constants with \
 * Make a clear distinction between is_file() calls and file_exists() calls
 * And a lot of other minors here and there

This patch supersedes https://github.com/Kruithne/KrameWork7/pull/85

### CODE REVIEW

**Comment by Kruithne** [comment]
I was going to contest this backslash madness on the basis of most core functions not providing their own ZVM opcodes, thus rendering this optimization pointless, but then I did some experimenting. The following example demonstrates the optimization I refer to.

```php
$foo = "This is a string of text.";
$length = strlen($foo);
```

When compiled down into opcodes for the ZVM, this produces the following output.

```function name: (null)
L1-3 {main}() C:\Program Files (x86)\PHP7\test.php - 0x2663070 + 4 ops
 L2    #0     ASSIGN                  $foo                 "This is a string "+
 L3    #1     STRLEN                  $foo                                      ~1
 L3    #2     ASSIGN                  $length              ~1
 L3    #3     RETURN                  1
```

The key optimization here is that the `strlen` function call is compiled down into the `STRLEN` opcode rather than an `FCALL`. Encapsulating the snippet in a namespace removes it since the namespace could define its own `strlen`.

```L1-4 {main}() C:\Program Files (x86)\PHP7\test.php - 0x2a72000 + 6 ops
 L3    #0     ASSIGN                  $foo                 "This is a string "+
 L4    #1     INIT_NS_FCALL_BY_NAME                        "MyNamespace\\strl"+
 L4    #2     SEND_VAR_EX             $foo                 1
 L4    #3     DO_FCALL                                                          @1
 L4    #4     ASSIGN                  $length              @1
 L4    #5     RETURN                  1
```

The reason I prefix the `strlen` function with the namespace operator, is to enforce the fact that I am not calling this function from the current namespace, restoring the opcode optimization.

```function name: (null)
L1-4 {main}() C:\Program Files (x86)\PHP7\test.php - 0x2672000 + 4 ops
 L3    #0     ASSIGN                  $foo                 "This is a string "+
 L4    #1     STRLEN                  $foo                                      ~1
 L4    #2     ASSIGN                  $length              ~1
 L4    #3     RETURN                  1
```

I looked over the functions you prefixed, and didn't find any custom opcodes at the ZVM level. But, then I noticed that instead of `DO_FCALL`, it actually uses `DO_ICALL` to indicate a code function.

```php
// Without namespace operator.
namespace MyNamespace {
	$foo = "This is a string of text.";
	$length = base64_encode($foo);
}
```
```function name: (null)
L1-5 {main}() C:\Program Files (x86)\PHP7\test.php - 0x28630c0 + 6 ops
 L4    #0     ASSIGN                  $foo                 "This is a string "+
 L5    #1     INIT_NS_FCALL_BY_NAME                        "MyNamespace\\base"+
 L5    #2     SEND_VAR_EX             $foo                 1
 L5    #3     DO_FCALL                                                          @1
 L5    #4     ASSIGN                  $length              @1
 L5    #5     RETURN
```
```php
// With namespace operator.
namespace MyNamespace {
	$foo = "This is a string of text.";
	$length = \base64_encode($foo);
}
```
```
function name: (null)
L1-6 {main}() C:\Program Files (x86)\PHP7\test.php - 0x2a630c0 + 6 ops
 L5    #0     ASSIGN                  $foo                 "This is a string "+
 L6    #1     INIT_FCALL              64                   "base64_encode"
 L6    #2     SEND_VAR                $foo                 1
 L6    #3     DO_ICALL                                                          @1
 L6    #4     ASSIGN                  $length              @1
 L6    #5     RETURN                  1
```

After some quick benchmarking, I was happy to see this *does* actually provide a very slight low-level optimization.

File: src/HTTP/HTTPHeader.php [hunk 1]
```diff
@@ -39,7 +39,7 @@
 		 * @api apply
 		 */
 		public function apply() {
-			header($this->__toString());
+			\header((string $this);
```
Comment (Kruithne) [inline src/HTTP/HTTPHeader.php]: Missing closing bracket to the string cast (?).

File: src/HTTP/HTTPContext.php [hunk 2]
```diff
@@ -45,14 +45,18 @@ class HTTPContext
 		 * @return string|null
 		 */
 		public static function getClientIP() {
-			if (!isset($_SERVER))
+			static $ip;
```
Comment (Kruithne) [inline src/HTTP/HTTPContext.php]: Keeping in consistency with the rest of the project, `$ip` should be defined as a static on the `HTTPContext` class and referred to as `self::$ip`. Unless there is some kind of low-level optimization behind this?

File: src/Storage/File.php [hunk 3]
```diff
@@ -118,11 +118,17 @@ public function getExtension():string {
 		 * @return string
 		 */
 		public function getFileType():string {
-			if (function_exists('finfo_file')) {
-				$info = new \finfo(FILEINFO_MIME_TYPE);
+			static $has_fileinfo;
```
Comment (Kruithne) [inline src/Storage/File.php]: Keeping in consistency with the rest of the project, `$has_fileinfo` should be `$hasFileInfo`, defined as a static on the `HTTPContext` class, and referred to as `self::$hasFileInfo`.

**Review by Kruithne — CHANGES_REQUESTED** [review]

File: src/Storage/File.php (hunk 3 above)
Comment (KalleZ) [reply src/Storage/File.php]: Fixed

File: src/HTTP/HTTPContext.php (hunk 2 above)
Comment (KalleZ) [reply src/HTTP/HTTPContext.php]: Fixed, I doubt there is a major gain by doing this, other than a property access call saved but hey they are not so expensive

File: src/HTTP/HTTPHeader.php (hunk 1 above)
Comment (KalleZ) [reply src/HTTP/HTTPHeader.php]: Fixed

**Comment by KalleZ** [comment]
Indeed it does do a slight optimization, the reason why is that internally PHP will chain check each namespace its currently in until it hits the global scope and since we know that these functions we want to use always live in the global scope, we might as well be nice and tell PHP that its where they are

**Review by Kruithne — APPROVED** [review]

**Comment by KalleZ** [comment]
As for the strlen() opcode, there are a few core builtin functiosn in PHP (as in the Zend Engine, not the standard library), that got special opcodes that can be used to optimize out some FCALLs:

 * count() (and sizeof())
 * in_array()
 * gettype()
 * get_class()
 * get_called_class()
 * func_num_args()
 * func_get_args()
 * strlen()
 * defined()

I'm not sure if pow() also had its own opcode after the introduction of the pow operator, but these are just what I remember on top of my head

**Comment by Kruithne** [comment]
Are you sure on that? I checked a few, I'm not seeing custom opcodes in the compilation for those calls. As an example, `count` is switching to `ICALL` with the namespace prefix, but does not have a custom opcode.
```function name: (null)
L1-5 {main}() C:\Program Files (x86)\PHP7\test.php - 0x2872000 + 6 ops
 L4    #0     ASSIGN                  $foo                 array(4)
 L5    #1     INIT_FCALL              64                   "count"
 L5    #2     SEND_VAR                $foo                 1
 L5    #3     DO_ICALL                                                          @1
 L5    #4     ASSIGN                  $length              @1
 L5    #5     RETURN                  1```

**Comment by KalleZ** [comment]
It may only be in more recent versions of PHP7, I haven't paid too much attention to some of the in depth changes to ZE for a while, but you can check their opcode handlers in php-src/Zend/zend_vm_execute.h:

http://git.php.net/?p=php-src.git;a=blob;f=Zend/zend_vm_execute.h;h=6b83684b7f3386afe097c19f6b9a8fd577a0e3c3;hb=HEAD

Where you can see the 4 opcode handlers for ZEND_COUNT (grep for ZEND_COUNT_SPEC_):

 * CONST
 * TMP
 * VAR
 * CV (Compiled Variable)

**Comment by Kruithne** [comment]
Hm, my environment is a few versions out-of-date; I'll need to fix that. Either way, since this PR puts the prefix on every global function we'll benefit from `ICALL` in the short-term, and any opcode optimizations they add in new (or future) versions will just be inherent.

**Comment by KalleZ** [comment]
Let me know if there is anything else for this PR :)

**Pull request merged**